### ISSUE TITLE
TITO strict gate trips on model-side behavior (alternate EOS, mid-turn control tokens); consumers must pass the family's parsers

### ISSUE THREAD

**Issue by nblintao** [radixark/miles#3113]
## Summary

Running the TITO strict gate (`--ci-test`) on a real agent harness — terminus-2 × Qwen3-0.6B, the Harbor GPU e2e in #2876 — trips it for three independent reasons. One is a consumer misconfiguration the gate is right to catch; two are model-side behaviors the comparator currently files as hard TITO bugs. Evidence comes from a 2026-09-09 run with `--dump-details`; recipe at the end.

## Mechanism 1 — the consumer did not pass the family's parsers (fixed in #2876)

Without `--sglang-reasoning-parser qwen3` the engine returns no `reasoning_content`; harbor's `Chat(interleaved_thinking=True)` only carries reasoning back into history when that field is non-empty, so the qwen3 template re-rendered an empty `<think></think>` skeleton against a recorded stream with real think tokens → `special_token_count` on every sample. Passing the parsers the session-verify lane uses (`qwen3` / `qwen25`) removes this class. The NeMo-Gym smoke's rate=1.0 (2026-07) most likely had the same cause.

Hardening idea: `TITOTokenizer` subclasses already declare `reasoning_parser` / `tool_call_parser`; a consumer that resolves a tito family could default the sglang flags from it instead of every launcher restating them.

## Mechanism 2 — the model ends some turns with `<|endoftext|>` (fix in progress)

Qwen3-0.6B's `generation_config.eos_token_id` is `[<|im_end|>, <|endoftext|>]` (the second is also its pad token) and the small model samples the second one for some turns. The template can only render `<|im_end|>\n`, and `Qwen3TITOTokenizer` assumes the `<|im_end|>` boundary — `merge_tokens` inserts the missing newline only after it — so the recorded stream and the canonical render disagree structurally:

```
expected: ... [<|im_start|>] (279 tokens) [<|im_end|>] (1 tokens) [<|im_start|>] ...
actual:   ... [<|im_start|>] (279 tokens) [<|endoftext|>]         [<|im_start|>] ...
```

Fix: #3161 (draft): the Qwen3 family treats either EOS as the turn end for the `\n` boundary, declares the pair as an EOS alias group, and the comparator reports an alias swap as a new non-severe type `eos_alias` (logged per type by `metrics.py`; the strict tuple is unchanged). Training tokens stay the recorded ones — only the verifier's judgement changes. Verified: Stage 2 `verify_chat_template.py --model Qwen/Qwen3-0.6B --tito-model qwen3 --thinking both` 74/74; targeted fast tests 189; Stage 3 healthy e2e session passes the strict gate (`assistant_text` ×2, `eos_alias` ×1). Large models in the session-verify lane (27–35B) rarely emit `<|endoftext|>` mid-chat, which is why that lane never saw this.

## Mechanism 3 — the model emits a raw `<|im_start|>` inside its own turn (open)

In a degenerate session (the model looped on terminus's XML format for ~48 turns until the engine's 40960-token context limit), the model emitted `<|im_start|>` as the first token of many of its turns:

```
recorded: [<|im_start|>] "assistant\n" [<|im_start|>] "\nOkay, the user's previous response had parsing errors ..." [<|im_end|>] "\n"
```

`_is_assistant_content` attributes a content segment by the immediately preceding special token, so those think segments are classified non-assistant, and their missing `</think>` closer lands in the hard `non_assistant_text` class (×33 in that sample). Structural mismatches masked this until mechanism 2 was fixed — the comparator returns only the structural record when segment counts differ.

The model's behavior cannot be controlled; what can change is how the comparator decides whose text a segment is. Two candidate approaches, both comparator-side:

- Attribute role from the session's generated spans instead of the nearest special token. The session already knows which tokens the model produced (v2 `TrajectoryNode.completion_span`; v1 can record the same per checkpoint), and a content segment that contains any generated token is assistant content. Family-agnostic, and it also handles a model-emitted `<|im_start|>user\n`. Needs both registries' `compute_session_mismatch` to pass the spans; the current heuristic stays as the fallback when none are given.
- Walk back to the turn's opener (`assistant_start_str`) instead of the nearest special. Comparator-only, but opener shapes differ per family and it cannot tell a model-emitted `<|im_start|>user\n` from a real user turn.

Either way the difference then lands in `assistant_text` and stays visible in the per-type rates. Consumer-side, #2876 moved to rollout-only on GLM-4.7-Flash (the recipe's model) with the strict gate armed and no opt-out flag: 2026-09-09, 2 terminus sessions of 9 and 30 turns, both reward 1.0, all four mismatch types 0.0. That leaves this mechanism a comparator-accuracy question for small or degenerate models, not a CI blocker.

## Open design questions (for the TITO / CI owners)

1. Is the strict gate meant to catch only TITO/template bugs? If so, are `eos_alias` (mechanism 2) and opener-based role attribution (mechanism 3) the intended way to keep model-side behavior visible without failing the gate?
2. Should the headline `tito_session_mismatch_rate` count only hard types, now that another non-severe type exists?
3. The session-verify lane runs only 27–35B models, so none of this surfaces there — should the TITO CI include a small-model lane, or is small-model behavior out of scope by design?

## Evidence recipe

Run the consumer with `--dump-details <dir>` (the rollout dump is written before the train-side assertion). `torch.load(<dir>/rollout_data/0.pt)` gives each sample's `metadata["tito_session_mismatch"]` (records carry `type`, `expected_text`, `actual_text`, `detail`) and `metadata["accumulated_token_ids"]` (the full session stream; the sample's own `tokens` are trimmed to max-seq-len). Re-segment the accumulated stream with `TokenSeqComparator.segment_by_special_tokens` to inspect a reported `segment_index`. Artifacts for this run: `/personal/tao.lin/tito-stage3-dump/` and `/personal/tao.lin/harbor-e2e-tito-stage3-20260909.log` on the sci cluster.

## Follow-ups

- Land #3161 (mechanism 2).
- Mechanism 3: a separate change by the TITO owner once the design questions above are answered.
- #2876 no longer carries the opt-out flag or its plumbing (rollout-only on GLM-4.7-Flash); nothing there waits on mechanism 3.

**Comment by LOGO127** [comment]
I tried to reduce this against current `main` with `Qwen/Qwen3-0.6B`, the bundled `qwen3_fixed.jinja`, and two sequential tool-call turns.

One detail seems important before changing the comparator: for this tokenizer, `<think>` / `</think>` are IDs `151667` / `151668`, both `special=False` and absent from `all_special_ids`. With the generated `<|im_end|>` retained, replacing the canonical empty skeleton with either no think block or a real think block produces only `assistant_text` mismatches on both turns. I can reproduce `special_token_count`, but only by removing `<|im_end|>` from the recorded completion; that already fails on turn 1, which does not match the reported single-turn-pass / multi-turn-fail pattern.

Could you share one serialized entry from `tito_session_mismatch` at the first failing turn (especially `detail`, `expected_text`, and `actual_text`), plus the SGLang commit/version used by the Harbor run? A short decode of the accumulated IDs around the first divergent boundary with `skip_special_tokens=False` would also be enough. That should distinguish a stop-token/logprob boundary issue from tokenizer metadata or a different historical-message shape, and let us keep any whitelist narrowly scoped instead of weakening the strict gate.

**Comment by nblintao** [comment]
Closing with the outcome recorded here. Mechanism 1 is fixed on the consumer side (pass the family's parsers). The consumer that hit mechanisms 2 and 3 moved to GLM-4.7-Flash rollout-only with the strict gate armed: 2026-09-09, two terminus sessions, both reward 1.0, all four mismatch types 0.0. The owner's read is that the remaining two mechanisms are small-model degeneration, outside what the strict gate is for; #3161 is closed on that basis. The evidence recipe and the two candidate comparator changes stay in this issue for reopening if a production model shows the same behavior.

**Comment by guapisolo** [comment]
codex cmt

To clarify Jiajun's conclusion: `<|endoftext|>` here is the pretraining/padding token surfacing in chat from weakly post-trained models, especially Qwen3-0.6B. He encountered this in March 2026 while building the first TITO version; it is much rarer, but still possible, in larger models. #3167 records that context in `Qwen3TITOTokenizer`.

Calling it model degeneration distinguishes its cause from a TITO implementation bug; it does not mean the strict gate should let it pass. In particular, do not treat it as an accepted `<|im_end|>` alias merely because generation can stop on it. The updated #2876 keeps the gate enabled with GLM-4.7-Flash, so the sandbox test no longer needs a comparator relaxation.
